tests/test_multislice.py

```diff
@@ -0,0 +1,365 @@
+
+    # pos1 = ax[1].imshow( np.abs( proj_dim1_Tvol3D ), cmap = 'turbo', )
+    # ax[1].set_title('sum1, proj along x (side view)')
+    # ax[1].set_xlabel("z-axis")
+    # ax[1].set_ylabel("y-axis")
+    # divider1 = make_axes_locatable(ax[1])
+    # cax1 = divider1.append_axes("right", size="5%", pad=0.1,)
+    # fig.colorbar(pos1, cax=cax1, fraction=0.046, pad=0.04)
+
+    # pos2 = ax[2].imshow( np.abs( proj_dim2_Tvol3D ), cmap = 'turbo', )
+    # ax[2].set_title('sum2, proj along z (beam)')
+    # ax[2].set_xlabel("x-axis")
+    # ax[2].set_ylabel("y-axis")
+    # divider2 = make_axes_locatable(ax[2])
+    # cax2 = divider2.append_axes("right", size="5%", pad=0.1,)
+    # fig.colorbar(pos2, cax=cax2, fraction=0.046, pad=0.04 )
+
+    # fig.tight_layout()
+    # plt.show(block=False)
+
+    # plt.close( fig )
+
+    #####
+
+    return Tvol3D
+
+#==================================================================================================
+# define probe function
+
+def create_2Dprobe_blurred_circ( array_size, circ_radius, blur_gauss_stdev ):
+
+    center = ( int( array_size[0] / 2 ), 
+               int( array_size[1] / 2 ))
+
+    X, Y = np.meshgrid( np.linspace( 0, array_size[1], array_size[1] ), 
+                        np.linspace( 0, array_size[0], array_size[0] ))
+
+    ellipse = ( (((( X - center[1]) / circ_radius[1] ) ) ** 2 ) + (((( Y - center[0]) / circ_radius[0] ) ) ** 2 )) <= 1
+    
+    ellipse_blurred = scipy.ndimage.gaussian_filter( input = np.abs( np.asarray( ellipse, dtype='cfloat')), 
+                                                     sigma = blur_gauss_stdev )
+    
+    return ellipse_blurred
+
+#==================================================================================================
+# define Fresnel integral spectrum propagation parameters
+
+def create_spectrum_propagator( N, wavelength, delta_z, prb_FOV ):
+
+    rr2 = np.linspace( -0.5 * N[0], 0.5 * N[0] - 1, num = N[0] ) ** 2
+    cc2 = np.linspace( -0.5 * N[1], 0.5 * N[1] - 1, num = N[1] ) ** 2
+
+    x = -1j * np.pi * wavelength * delta_z
+    rr2 = np.exp( x * rr2[ ..., None ] / ( prb_FOV[0] ** 2 ))
+    cc2 = np.exp( x * cc2[ ..., None ] / ( prb_FOV[1] ** 2 ))
+
+    spect_prop_phase_curvature = np.fft.fftshift( np.outer( np.transpose( rr2 ), cc2 ))
+
+    return spect_prop_phase_curvature
+
+#==================================================================================================
+
+def forward_multislice( obj3D, phi, spect_prop_phase_curvature, ):
+
+    exwv3D = cp.zeros( obj3D.shape, dtype = obj3D.dtype )
+    prb3D  = cp.zeros( obj3D.shape, dtype = obj3D.dtype )
+
+    for tt in cp.arange( 0, obj3D.shape[-1], 1 ) :
+
+        prb3D[  :, :, tt ] = phi
+        exwv3D[ :, :, tt ] = obj3D[ :, :, tt ] * phi
+
+        if tt == obj3D.shape[-1]:
+            break
+
+        phi = cp.fft.ifft2( cp.fft.fft2( exwv3D[ :, :, tt ] ) * spect_prop_phase_curvature )
+
+    return exwv3D, prb3D
```

Comment:
It looks like these two function are what define the propagation function which can propagate a wave between two multi-slice planes. These functions need

- [ ] documentation and type hints which explain the input and output parameters
- [ ] explanatory text which identify which algorithm/physics model the functions implement
- [ ] possibly citations to any manuscripts which describe the implementation if the implementation is not obvious